From d2f6802938c2cdcd18ddab5ab7171112368d4701 Mon Sep 17 00:00:00 2001
From: Botond Kalocsai <kalocsaibotond@gmail.com>
Date: Thu, 26 Sep 2024 21:05:18 +0200
Subject: [PATCH] feat: Add character rendering font size proportional offset
   for correcting different font dimensions after wide glyph support patch.

---
 config.def.h |  3 +++
 x.c          | 12 ++++++++----
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/config.def.h b/config.def.h
index 2cd740a..66f8c24 100644
--- a/config.def.h
+++ b/config.def.h
@@ -28,6 +28,9 @@ char *vtiden = "\033[?6c";
 /* Kerning / character bounding-box multipliers */
 static float cwscale = 1.0;
 static float chscale = 1.0;
+/* Character rendering font size proportional offset scale values. */
+static float cxpropoffset = 0;
+static float cypropoffset = 0;
 
 /*
  * word delimiter string
diff --git a/x.c b/x.c
index b093df0..2e5cb55 100644
--- a/x.c
+++ b/x.c
@@ -1291,8 +1291,8 @@ xmakeglyphfontspecs(XftGlyphFontSpec *specs, const Glyph *glyphs, int len, int x
 		if (glyphidx) {
 			specs[numspecs].font = font->match;
 			specs[numspecs].glyph = glyphidx;
-			specs[numspecs].x = (short)xp;
-			specs[numspecs].y = (short)yp;
+			specs[numspecs].x = (short)round(xp + cxpropoffset * usedfontsize);
+			specs[numspecs].y = (short)round(yp + cypropoffset * usedfontsize);
 			xp += runewidth;
 			numspecs++;
 			continue;
@@ -1491,12 +1491,16 @@ xdrawglyphfontspecs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, i
 
         /* Render underline and strikethrough. */
         if (base.mode & ATTR_UNDERLINE) {
-            XftDrawRect(xw.draw, fg, winx, winy + dc.font.ascent + 1,
+            XftDrawRect(xw.draw, fg, winx,
+                        round(cypropoffset * usedfontsize) +
+                        winy + dc.font.ascent + 1,
                         width, 1);
         }
 
         if (base.mode & ATTR_STRUCK) {
-            XftDrawRect(xw.draw, fg, winx, winy + 2 * dc.font.ascent / 3,
+            XftDrawRect(xw.draw, fg, winx,
+                        round(cypropoffset * usedfontsize) +
+                        winy + 2 * dc.font.ascent / 3,
                         width, 1);
         }
     }
-- 
2.46.1.windows.1

